### PR TITLE
SupCon loss

### PR DESCRIPTION
**Pull request by fjsj**
Based on https://github.com/HobbitLong/SupContrast/blob/master/losses.py
Closes https://github.com/KevinMusgrave/pytorch-metric-learning/issues/281

### CODE REVIEW

File: tests/losses/test_supcon_loss.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                        )
+                    anchor_lossA += torch.log(numeratorA / denominatorA)
+                    anchor_lossB += torch.log(numeratorB / denominatorB)
+                total_lossA += -anchor_lossA / len(pos_pairs)
+                total_lossB += -anchor_lossB / len(pos_pairs)
+            total_lossA /= len(anchor_pos_pairs)
+            total_lossB /= len(anchor_pos_pairs)
+            rtol = 1e-2 if dtype == torch.float16 else 1e-5
+            self.assertTrue(torch.isclose(lossA, total_lossA, rtol=rtol))
+            self.assertTrue(torch.isclose(lossB, total_lossB, rtol=rtol))
```
Comment (fjsj) [inline tests/losses/test_supcon_loss.py:94]: I already had implemented this before I saw your recent changes on `tests/losses/test_ntxent_loss.py`. Should we change this to match it?

File: src/pytorch_metric_learning/losses/supcon_loss.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+from ..distances import CosineSimilarity
+from ..reducers import AvgNonZeroReducer
+from ..utils import common_functions as c_f
+from ..utils import loss_and_miner_utils as lmu
+from .generic_pair_loss import GenericPairLoss
+
+
+class SupConLoss(GenericPairLoss):
+    def __init__(self, temperature=0.1, **kwargs):
```
Comment (fjsj) [inline src/pytorch_metric_learning/losses/supcon_loss.py:9]: I removed the `base_temperature` parameter from https://github.com/HobbitLong/SupContrast/blob/master/losses.py

This is not mentioned in the paper, and it's useless if it's set to be equal to temperature.

**Comment by KevinMusgrave** [comment]
@fjsj I'd like to make some changes. Can you allow edits on your fork? See this: https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

**Comment by fjsj** [comment]
Hi @KevinMusgrave , I always have trouble with that option... it's not showing to me. I've added you as a collaborator to the https://github.com/vintasoftware/pytorch-metric-learning/ project. Please LMK if this works.

**Comment by KevinMusgrave** [comment]
Thanks that worked. The main change I made was merging ```test_supcon_loss``` with ```test_ntxent_loss``` since they are so similar. In the process of doing that I realized that NTXentLoss + PerAnchorReducer actually isn't equivalent to SupConLoss. The difference is:

- NTXentLoss: the denominator contains 1 positive pair (from the numerator) and all negative pairs
- SupConLoss: the denominator contains all positive and negative pairs
